### PR TITLE
Feature: prometheus metrics

### PR DESCRIPTION
**Pull request by rjonczy**
### Server
`http` server exposing `/metrics` endpoint is starting on a separate go routine on separe http server (`:2112`).

### Metrics

```
> $ curl -s http://localhost:2112/metrics | grep 'taraxa_indexer'    
# HELP taraxa_indexer_average_block_processing_time Average time of processing block in milisec
# TYPE taraxa_indexer_average_block_processing_time gauge
taraxa_indexer_average_block_processing_time 8
# HELP taraxa_indexer_blocks_indexed_counter Number of indexed blocks by indexer since restart
# TYPE taraxa_indexer_blocks_indexed_counter counter
taraxa_indexer_blocks_indexed_counter 298
# HELP taraxa_indexer_blocks_indexed_total Total number of indexed blocks
# TYPE taraxa_indexer_blocks_indexed_total gauge
taraxa_indexer_blocks_indexed_total 32504
# HELP taraxa_indexer_dag_blocks_indexed_total Total number of indexed DAG blocks
# TYPE taraxa_indexer_dag_blocks_indexed_total gauge
taraxa_indexer_dag_blocks_indexed_total 208225
# HELP taraxa_indexer_pbft_blocks_indexed_total Total number of indexed PBFT blocks
# TYPE taraxa_indexer_pbft_blocks_indexed_total gauge
taraxa_indexer_pbft_blocks_indexed_total 32206
# HELP taraxa_indexer_rpc_calls_counter The total number of RPC calls to taraxa-node
# TYPE taraxa_indexer_rpc_calls_counter counter
taraxa_indexer_rpc_calls_counter 300
# HELP taraxa_indexer_storage_add_counter The total number of Add operations to storage
# TYPE taraxa_indexer_storage_add_counter counter
taraxa_indexer_storage_add_counter 0
# HELP taraxa_indexer_storage_get_counter The total number of Get operations to storage
# TYPE taraxa_indexer_storage_get_counter counter
taraxa_indexer_storage_get_counter 2556
# HELP taraxa_indexer_transactions_indexed_total Total number of indexed transactions
# TYPE taraxa_indexer_transactions_indexed_total gauge
taraxa_indexer_transactions_indexed_total 1.901186e+06
```

Commit log:
- initial prometheus metrics
- rename metrics
- added metricsPrefix
- added more metrics
- added another metric

### CODE REVIEW

File: internal/indexer/indexer.go [hunk 1]
```diff
@@ -111,6 +115,17 @@ func (i *Indexer) sync() (err error) {
...
 		}
+
+		counter++
+
+		// prometheus metics
+		metrics.IndexedBlocksCounter.Inc()
+		metrics.AverageBlockProcessingTimeMilisec.Set(float64((time.Since(prev).Milliseconds() / counter)))
+		metrics.IndexedBlocksTotal.Set(float64(chain.ParseInt(blk.Number)))
+		metrics.IndexedDagBlocksTotal.Set(float64(finalizationData.DagCount))
+		metrics.IndexedPbftBlocksTotal.Set(float64(finalizationData.PbftCount))
```
Comment (kstdl) [inline internal/indexer/indexer.go:126]: Looks like it is should be the same thing as `IndexedBlocksTotal`
This code will set just the same data here every time as you getting `finalizationData` from db before this for cycle. Should be:
```
		metrics.IndexedPbftBlocksTotal.Set(float64(p))
		metrics.IndexedDagBlocksTotal.Set(float64(dc))
		metrics.IndexedTransactionsTotal.Set(float64(tc))
```
Also this `AverageBlockProcessingTimeMilisec` isn't set correctly. As `prev` is reset every 100 blocks. So you should set counter near [this line](https://github.com/Taraxa-project/taraxa-indexer/pull/54/files#diff-87c2475baa87005bd3a3001ec33f201feea23caddbbf5a2f249238a2bd32026dL107) also to get correct data here

Comment (kstdl) [reply internal/indexer/indexer.go:126]: And also we are processing blocks not only during syncing. So you need to set all this data here https://github.com/Taraxa-project/taraxa-indexer/pull/54/files#diff-87c2475baa87005bd3a3001ec33f201feea23caddbbf5a2f249238a2bd32026dR173
Or maybe even better will be just to move it inside of this `process` method

File: internal/chain/ws_client.go [hunk 2]
```diff
@@ -27,6 +29,7 @@ func NewWsClient(url string) (*WsClient, error) {
 func (client *WsClient) GetBlockByNumber(number uint64) (blk *Block, err error) {
 	blk = new(Block)
 	err = client.rpc.Call(blk, "eth_getBlockByNumber", fmt.Sprintf("0x%x", number), false)
+	metrics.RpcCallsCounter.Inc()
```
Comment (kstdl) [inline internal/chain/ws_client.go:32]: Right now it looks like `GetBlockByNumberCounter`. You should add this to all RPC methods I think

Comment (rjonczy) [reply internal/chain/ws_client.go:32]: Yeah. That was an intention of this metric, to capture all rpc calls. Just forgot to add to  all methods. Fixed now.

**Comment by rjonczy** [comment]
@kstdl - redesign mertrics, and added more of them.

Currenctly following metrics and exposed.

```
# HELP taraxa_indexer_block_processing_time Time of processing block in milisec
# TYPE taraxa_indexer_block_processing_time summary
taraxa_indexer_block_processing_time_sum 113328
taraxa_indexer_block_processing_time_count 230
# HELP taraxa_indexer_blocks_indexed_counter Number of indexed blocks by indexer since start
# TYPE taraxa_indexer_blocks_indexed_counter counter
taraxa_indexer_blocks_indexed_counter 230
# HELP taraxa_indexer_blocks_indexed_total Total number of indexed blocks
# TYPE taraxa_indexer_blocks_indexed_total gauge
taraxa_indexer_blocks_indexed_total 3929
# HELP taraxa_indexer_dags_indexed_total Total number of indexed DAG blocks
# TYPE taraxa_indexer_dags_indexed_total gauge
taraxa_indexer_dags_indexed_total 26389
# HELP taraxa_indexer_indexed_dags Number of indexed DAG blocks since start
# TYPE taraxa_indexer_indexed_dags gauge
taraxa_indexer_indexed_dags 1421
# HELP taraxa_indexer_indexed_dags_last_block Number of indexed DAG blocks in last processed block
# TYPE taraxa_indexer_indexed_dags_last_block gauge
taraxa_indexer_indexed_dags_last_block 8
# HELP taraxa_indexer_indexed_transactions Number of indexed transactions since start
# TYPE taraxa_indexer_indexed_transactions gauge
taraxa_indexer_indexed_transactions 13279
# HELP taraxa_indexer_indexed_transactions_last_block Number of indexed transactions in last processed blo
ck
# TYPE taraxa_indexer_indexed_transactions_last_block gauge
taraxa_indexer_indexed_transactions_last_block 67
# HELP taraxa_indexer_rpc_calls_counter The total number of RPC calls to taraxa-node
# TYPE taraxa_indexer_rpc_calls_counter counter
taraxa_indexer_rpc_calls_counter 28672
# HELP taraxa_indexer_storage_commit_counter The total number of commit operations to storage
# TYPE taraxa_indexer_storage_commit_counter counter
taraxa_indexer_storage_commit_counter 230
# HELP taraxa_indexer_time_last_block Timestamp of last processed block
# TYPE taraxa_indexer_time_last_block gauge
taraxa_indexer_time_last_block 1.6789620526254272e+09
# HELP taraxa_indexer_transactions_indexed_total Total number of indexed transactions
# TYPE taraxa_indexer_transactions_indexed_total gauge
taraxa_indexer_transactions_indexed_total 272240
```

**Review by kstdl — APPROVED** [review]

**Pull request merged**